contracts/token/non-fungible/src/storage.rs

```diff
@@ -0,0 +1,357 @@
+        panic_with_error!(e, NonFungibleTokenError::InvalidLiveUntilLedger);
+    }
+
+    let key = StorageKey::Approval(token_id);
+
+    let approval_data = ApprovalData { approver: approver.clone(), live_until_ledger };
+
+    e.storage().temporary().set(&key, &approval_data);
+
+    let live_for = live_until_ledger - e.ledger().sequence();
+
+    e.storage().temporary().extend_ttl(&key, live_for, live_for);
+
+    emit_approval(e, owner, approver, token_id, live_until_ledger);
+}
+
+/// Sets or removes operator approval for managing all tokens owned by the
+/// owner.
+///
+/// # Arguments
+///
+/// * `e` - Access to the Soroban environment.
+/// * `owner` - The address granting approval for all their tokens.
+/// * `operator` - The address being granted or revoked approval.
+/// * `approved` - If true, grants approval; if false, revokes approval.
+/// * `live_until_ledger` - The ledger number at which the approval expires.
+///
+/// # Errors
+///
+/// * [`NonFungibleTokenError::InvalidLiveUntilLedger`] - If the ledger number
+///   is less than the current ledger number.
+pub fn set_approval_for_all(
+    e: &Env,
+    owner: &Address,
+    operator: &Address,
+    approved: bool,
+    live_until_ledger: u32,
+) {
+    owner.require_auth();
+
+    if live_until_ledger < e.ledger().sequence() {
+        panic_with_error!(e, NonFungibleTokenError::InvalidLiveUntilLedger);
+    }
+
+    let key = StorageKey::ApprovalForAll(owner.clone());
+
+    let approval_data =
+        ApprovalForAllData { operator: operator.clone(), approved, live_until_ledger };
+
+    e.storage().temporary().set(&key, &approval_data);
+
+    let live_for = live_until_ledger - e.ledger().sequence();
+
+    e.storage().temporary().extend_ttl(&key, live_for, live_for);
+
+    emit_approval_for_all(e, owner, operator, approved, live_until_ledger);
+}
+
+/// Low-level function for handling transfers for NFTs, but doesn't
+/// handle authorization. Updates ownership records, adjusts balances,
+/// and clears existing approvals.
+///
+/// # Arguments
+///
+/// * `e` - Access to the Soroban environment.
+/// * `from` - The address of the current token owner.
+/// * `to` - The address of the token recipient.
+/// * `token_id` - The identifier of the token being transferred.
+///
+/// # Errors
+///
+/// * [`NonFungibleTokenError::IncorrectOwner`] - If the `from` address is not
+///   the owner of the token.
+/// * [`NonFungibleTokenError::MathOverflow`] - If the balance of the `to` would
+///   overflow.
+pub fn update(e: &Env, from: Option<&Address>, to: Option<&Address>, token_id: u128) {
+    if let Some(from_address) = from {
+        let owner = owner_of(e, token_id);
+
+        // Ensure the `from` address is indeed the owner.
+        if owner != *from_address {
+            panic_with_error!(e, NonFungibleTokenError::IncorrectOwner);
+        }
+
+        // Update the balance of the `from` address
+        let from_balance = balance(e, from_address) - 1;
```

Comment:
do we need to check overflow as below?

Reply:
if you mean underflow, I don't think we need to do that. Because, if the `from_address` is not the owner, it will not even reach that balance reduction. And the current owner cannot have `0` balance.

Reply:
makes sense 👍 could be good to mention that in a comment

Reply:
good suggestion, will do 👍